RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationAlgorithms.dev.cc

```diff
@@ -5,6 +5,60 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
 
   using namespace cms::alpakatools;
 
+  //
+  struct HGCalRecHitCalibrationKernel_flagRecHits {
+    template <typename TAcc>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc, HGCalDigiDevice::View digis, HGCalRecHitDevice::View recHits) const {
+      for (auto idx : uniform_elements(acc, digis.metadata().size())) {
+        recHits[idx].flags() = digis[idx].flags();
+      }
+    }
+  };
+
+  //
+  struct HGCalRecHitCalibrationKernel_adcToCharge {
+    template <typename TAcc>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc, HGCalDigiDevice::View digis, HGCalRecHitDevice::View recHits, HGCalCalibParamDevice::ConstView calibs) const {
+
+      auto adc_to_fC = [&](uint32_t adc, uint32_t cm, uint32_t adcm1, float adc_ped, float cm_slope, float cm_ped, float bxm1_slope, float adc2fC) {
+        return adc2fC*( (adc-adc_ped) + cm_slope*(cm-cm_ped) + bxm1_slope*(adcm1-adc_ped) );
+      };
+      
+      auto tot_to_fC = [&](uint32_t tot, float tot_lin, float tot_ped, float tot2fC, float tot_p0, float tot_p1, float tot_p2) {
+        bool isLin(tot>tot_lin);
+        bool isNotLin(!isLin);
+        return isLin*( tot2fC*(tot-tot_ped) ) + isNotLin*( tot_p0 + tot_p1*tot + tot_p2*tot*tot);
+      };
+
+      for (auto idx : uniform_elements(acc, digis.metadata().size())) {
+
+        auto calib = calibs[idx];
+        auto digi = digis[idx];
+        bool useTOT(digi.tctp()==3);
+        bool useADC(!useTOT);
+        recHits[idx].energy() = \
+          useADC*adc_to_fC(digi.adc(), digi.cm(), digi.adcm1(), calib.ADC_ped(), calib.CM_slope(), calib.CM_ped(), calib.BXm1_slope(), calib.ADCtofC()) + \
+          useTOT*tot_to_fC(digi.tot(), calib.TOT_lin(), calib.TOTtofC(), calib.TOT_ped(), calib.TOT_P0(), calib.TOT_P1(), calib.TOT_P2());
```

Comment:
For my education: Is there a computational reason we are not using an if-statement here and in `tot_to_fC` ?

In the case of `useADC*adc_to_fC(...) + useTOT*tot_to_fC(...)`, won't both functions be processed no matter the value of the booleans `useADC` and `useTOT`, at a small performance cost per evaluation?

Apologies if I am guilty of premature optimization. :p

Reply:
I should maybe test if we see any difference. I understood from the tutorials that if-else statements should be avoided as they lead to a branching of the usage of the cores which may ended up being used somehow sequentially. I found this also in stackoverflow.

https://stackoverflow.com/questions/45734138/why-an-if-else-statement-in-gpus-code-will-cut-the-performance-in-half

Reply:
Hi @pfs, okay thanks, I indeed thought something like this was the case. I remember @adavidzh mentioned branchless programming: https://en.algorithmica.org/hpc/pipelining/branchless/ but I have to study it a bit deeper. We can also ask feedback from the Alpaka experts, they probably know how this is also optimized by the compiler.